File: typescript/src/calisthenics-exercises/not-more-than-2-instance-variables.ts
export class CheckoutSession {
  private cartItems: Array<{ id: string; price: number; qty: number }> = []
  private customerId: string | null = null
  private shippingAddress: string | null = null
  private billingAddress: string | null = null
  private couponCode: string | null = null
  private paymentMethod: 'CARD' | 'PAYPAL' | null = null
  private currency: string = 'USD'
  private taxRate: number = 0.21

  addItem(id: string, price: number, qty: number) {
    this.cartItems.push({id, price, qty})
  }

  total(): number {
    const subtotal = this.cartItems.reduce((sum, i) => sum + i.price * i.qty, 0)
    const discount = this.couponCode ? 10 : 0 // lógica de descuento primitiva
    const taxed = (subtotal - discount) * (1 + this.taxRate)
    return this.currency === 'USD' ? taxed : taxed * 0.9 // conversión simulada
  }
}

File: csharp/src/calisthenics-exercises/NotMoreThan2InstanceVariables.cs
namespace RefactoringAvanzado.CalisthenicExercises;

public class CartItem
{
    public string Id { get; set; }
    public double Price { get; set; }
    public int Qty { get; set; }

    public CartItem(string id, double price, int qty)
    {
        Id = id;
        Price = price;
        Qty = qty;
    }
}

public class CheckoutSession
{
    private List<CartItem> cartItems = new List<CartItem>();
    private string? customerId = null;
    private string? shippingAddress = null;
    private string? billingAddress = null;
    private string? couponCode = null;
    private string? paymentMethod = null;
    private string currency = "USD";
    private double taxRate = 0.21;

    public void AddItem(string id, double price, int qty)
    {
        cartItems.Add(new CartItem(id, price, qty));
    }

    public double Total()
    {
        var subtotal = cartItems.Sum(i => i.Price * i.Qty);
        var discount = couponCode != null ? 10 : 0;
        var taxed = (subtotal - discount) * (1 + taxRate);
        return currency == "USD" ? taxed : taxed * 0.9;
    }
}
